Encryption.md

```diff
@@ -132,49 +203,164 @@ the encryption buffer is comprised of an IV, ciphertext and tag, as described in
 
+```c
+struct ColumnChunk {
+...
+  /** Byte offset in file_path to the ColumnMetaData **/
+  2: required i64 file_offset
+
+  /** Column metadata for this chunk. This is the same content as what is at
+   * file_path/file_offset.  Having it here has it replicated in the file
+   * metadata.
+   **/
+  3: optional ColumnMetaData meta_data
+..
+}
+```
+
 ### Encrypted footer mode
 
 In files with sensitive column data, a good security practice is to encrypt not only the 
-secret columns, but also the file footer metadata, with a separate footer key. This hides
+secret columns, but also the file footer metadata with a separate footer key. This hides
 the file schema / column names, number of rows, key-value properties, column sort order, 
 column data offset and size, list of encrypted columns and metadata of the column encryption keys. 
-It also makes the footer tamper-proof.
+It also makes the footer tamper-proof and helps to prevent module replacement attacks.
 
-The columns encrypted with the same key as the footer, don't split the ColumnMetaData from the 
+The columns encrypted with the same key as the footer don't separate the ColumnMetaData from the 
 ColumnChunks, leaving it at the original location, `optional ColumnMetaData meta_data`. This field
-is not set for columns enrypted with a column-specific key.
+is not set for columns encrypted with a column-specific key.
 
 A Thrift-serialized `FileCryptoMetaData` structure is written after the footer. It contains 
 information on the file encryption algorithm and on the footer (offset in 
 the file, and an optional key metadata, with a maximal length of 512). Then 
-the length of this structure is written, as a 4-byte little endian integer. Then a final 
+the length of this structure is written as a 4-byte little endian integer, followed by a final 
 magic string, "PARE".
-
-Only the `FileCryptoMetaData` is written as a plaintext, all other file parts are protected
+The same magic bytes are written at the beginning of the file (offset 0).
+Parquet readers start file parsing by reading and checking the magic string. Therefore, the 
+encrypted footer mode uses a new magic string ("PARE") in order to  instruct readers to look 
+for a file crypto metadata at the end of the file - and also to immediately inform 
+legacy readers (expecting ‘PAR1’ bytes) that they can’t parse this file.
+
+```c
+/** Crypto metadata for files with encrypted footer **/
+struct FileCryptoMetaData {
+  /** 
+   * Encryption algorithm. Note that this field is only used for files
+   * with encrypted footer. Files with plaintext footer store the algorithm id
+   * inside footer (FileMetaData structure).
+   */
+  1: required EncryptionAlgorithm encryption_algorithm
+    
+  /** Retrieval metadata of key used for encryption of footer, 
+   *  and (possibly) columns **/
+  2: optional binary footer_key_metadata
+
+  /** Offset of encrypted Parquet footer **/
+  3: required i64 footer_offset
```

Comment:
The offset at the end of the file should be the footer offset, just like in unencrypted files. If the FileCryptoMetaData is written after the footer, writing the footer offset there still works.

The reason for this is that we want to avoid backward seeks, which are expensive. If the footer offset is only available here, then there are 3 backward seeks: one to the offset and magic bytes, one to the start of FileCryptoMetaData, and one to the start of the Footer . . . and probably one to the start of the ColumnMetaDatas and maybe yet another to the start of the offset indexes!

Keeping the footer offset at the end before the magic bytes allows reading both the footer and the crypto metadata in one read and avoids an expensive backward seek.

Reply:
Good point, I haven't thought about seek costs. Actually, the current mr implementation also doesn't minimize the seeks - there is one to len-4 for magic bytes, one to len-8 for footer length, and then to the footer. Its possible of course to seek len-8 and read the 8 bytes. I'll implement seek minimization for both plaintext and encrypted footers. In order to save the magic number seek, this will require a use of the current 4 byte field (shift) instead of 8 bytes (offset) - the footer offset is calculated as len - 8 - shift.
Shift is equal to the FileMetaData length for plaintext footers, and to the combined length of FileMetaData and FileCryptoMetaData for encrypted footers.

Reply:
I don't get what you're talking about using shift instead of offset. Either way, we have to store a value that gets us to the start of the footer. We should be using a fixed-width value there so that we can seek directly to the start of it to read the offset and magic at the same time.

I think that encrypted files should use the same method to get to the start of the footer that is used by unencrypted files. The start of the footer should be where the encrypted buffer's length or IV is stored, depending on which comes first.

Reply:
I'm saying there is no a footer offset field today (offset from the file start, an 8 byte long). Instead, there is a 4-byte int field that is used to calculate the footer offset, by backshifting the current position. Anyways, this is semantics. Your second comment says we are in sync on this.